### PR TITLE
120 - icon

### PR DESCRIPTION
**Pull request by RajaRamu**
## Description
DS6 icons

## References
Design - https://app.zeplin.io/project/5a742fd1f51a76f8651f2016/screen/5ac29323882d1f0d438f51ab
Issue - https://github.com/eBay/skin/issues/120

## Screenshots
screen shot 2018-04-09 at 9 20 49 pm
screen shot 2018-04-09 at 9 21 02 pm

### CODE REVIEW

**Comment by RajaRamu** [comment]
I have added only arrow background icons. I just want to get everyone's feedback before making rest of the icons. please help reviewing it.

File: docs/_includes/ds6/notice.html [hunk 1]
```diff
@@ -81,7 +81,7 @@ <h4 class="page-notice__status" id="confirmation-status">
                         <p>You can check <a href="#">your listing</a> or <a href="#">sell another item</a></p>
                     </div>
                     <button aria-label="Dismiss notice" class="page-notice__close">
-                        <span></span>
+                        <span class='icon icon--close'></span>
```
Comment (ianmcburnie) [inline docs/_includes/ds6/notice.html]: I don't think we should change the API/markup for notice module. Because the notice icon is always an X there is no need to require the developer to specify the icon, we abstract that detail away like we were doing currently.

File: src/less/notice/ds6/notice-mixin.less [hunk 2]
```diff
@@ -1,3 +1,4 @@
 .notice-mixin() {
+    @import "icon.less";
```
Comment (seangates) [inline src/less/notice/ds6/notice-mixin.less]: Why was this added? Icon's have their own mixin.

File: dist/less/ds6/mixins.less [hunk 3]
```diff
@@ -89,16 +89,3 @@
     .ds6-typography-small;
     color: @ds6-color-g206-gray;
 }
-
-.background-icon-base() {
```
[the remark this answers is not in the document]
Comment (RajaRamu) [reply dist/less/ds6/mixins.less]: it got moved to `src/less/less/mixins-common.less`.  there are lot of mixins from this file are used into ds6. i thought `background-icon-base` is common.

[the remark this answers is not in the document]
Comment (ianmcburnie) [reply dist/less/ds6/mixins.less]: `background-icon-base` was created specifically for DS6.

File: src/less/notice/ds6/notice-mixin.less (hunk 2 above)
Comment (ianmcburnie) [reply src/less/notice/ds6/notice-mixin.less]: Yes, this change looks invalid. Please remove.

**Comment by ianmcburnie** [comment]
@RajaRamu Could you also add the 4 new icons to the inline SVG example section:

screen shot 2018-04-06 at 1 29 49 pm

**Comment by ianmcburnie** [comment]
@RajaRamu There are a lot of conflicts. Can you try rebasing with `4.0.0` then running `yarn`, then pushing.

**Comment by ianmcburnie** [comment]
@RajaRamu Can you delete the old existing ds6 icons that are there. I see four in the inline SVG and one in background svg. Your new icons will supersede these.

The button module will need to be updated to use the new arrow right inline SVG in the CTA button. I don't think the other 3 inline SVGs were actually being used anywhere in skin.

**Comment by RajaRamu** [comment]
@ianmcburnie - i do not see any reference for these icons neither in button module nor any other module. can you please give me the pointer where its being used?
```
svg.icon--chevron-dropdown
svg.icon--arrow-cta
svg.icon--hamburger
svg.icon--close 
span.icon--close

```

**Comment by ianmcburnie** [comment]
@RajaRamu It's being used here:

screen shot 2018-04-11 at 3 08 00 pm

File: dist/less/ds6/mixins.less [hunk 4]
```diff
@@ -111,3 +111,13 @@
         .background-svg(@ds6-color-link-default);
     }
 }
+
```
Comment (ianmcburnie) [inline dist/less/ds6/mixins.less]: @RajaRamu Shouldn't the icon mixins be in here? e.g. `icon-close-mixin`? Like we have in ds4? Or maybe I'm misunderstanding how these icons will be used. Can you perhaps give an example of how these changes would be integrated into current ds6 page notice?

**Comment by RajaRamu** [comment]
@ianmcburnie - Based on our discussion, entire `ebay/skin/icon` will be loaded into ebayui-core. Only inline svg `<symbol>` will be aggregated and delivered as inline style dynamically using `marko/transformer`. Usage would be,
1. `<ebay-icon type='background' icon='icon-arrow-down aria-label='arrow down'/>` and output would be    `<span class="icon icon--arrow-down"></span>`.
2. `<ebay-icon type='inline' icon='icon-arrow-down aria-label='arrow down'/>` and output would be `<symbol id="icon-arrow-down" viewBox="0 0 22 22">
        <path d="M7 6.072L4.834 3.906a.656.656 0 0 0-.928.928L6.072 7 3.906 9.166a.656.656 0 0 0 .928.928L7 7.928l2.166 2.166a.656.656 0 0 0 .928-.928L7.928 7l2.166-2.166a.656.656 0 0 0-.928-.928L7 6.072zM7 14A7 7 0 1 1 7 0a7 7 0 0 1 0 14z"/>
    </symbol>` 
&&
 `<svg aria-hidden="true" class="icon icon--arrow-down" focusable="false" height="24" width="24">
                <use xlink:href="#icon-arrow-down"></use>
            </svg>`

`<symbol>` and `<svg>` tags  are in `ds6/icon.html` and `/ds6/svg.html`. i was thinking to extract into some less file so that it could be imported into `ebayui-core`. Contrary to that, `<symbol>` and `<svg>`  could be copied over which is not ideal. thought?

File: dist/less/ds6/mixins.less (hunk 4 above)
Comment (RajaRamu) [reply dist/less/ds6/mixins.less]: @ianmcburnie - Based on our discussion, entire `ebay/skin/icon` will be loaded into ebayui-core. Only inline svg `<symbol>` will be aggregated and delivered as inline style dynamically using `marko/transformer`. Usage would be,
1. `<ebay-icon type='background' icon='icon-arrow-down aria-label='arrow down'/>` and output would be    `<span class="icon icon--arrow-down"></span>`.
2. `<ebay-icon type='inline' icon='icon-arrow-down aria-label='arrow down'/>` and output would be `<symbol id="icon-arrow-down" viewBox="0 0 22 22">
        <path d="M7 6.072L4.834 3.906a.656.656 0 0 0-.928.928L6.072 7 3.906 9.166a.656.656 0 0 0 .928.928L7 7.928l2.166 2.166a.656.656 0 0 0 .928-.928L7.928 7l2.166-2.166a.656.656 0 0 0-.928-.928L7 6.072zM7 14A7 7 0 1 1 7 0a7 7 0 0 1 0 14z"/>
    </symbol>` 
&&
 `<svg aria-hidden="true" class="icon icon--arrow-down" focusable="false" height="24" width="24">
                <use xlink:href="#icon-arrow-down"></use>
            </svg>`

`<symbol>` and `<svg>` tags  are in `ds6/icon.html` and `/ds6/svg.html`. i was thinking to extract into some less file so that it could be imported into `ebayui-core`. Contrary to that, `<symbol>` and `<svg>`  could be copied over which is not ideal. thought?

**Comment by RajaRamu** [comment]
@ianmcburnie - i have removed unused icons. https://github.com/eBay/skin/pull/124/commits/c453b8dbbd8e24763ebea4797cf3154349024dd6. please continue to review
svg.icon--chevron-dropdown
svg.icon--hamburger
svg.icon--close 
span.icon--close

File: dist/less/ds6/mixins.less (hunk 4 above)
Comment (ianmcburnie) [reply dist/less/ds6/mixins.less]: @RajaRamu I'm not talking about ebayui-core. I'm talking purely from a skin perspective. How would end-users integrate an icon right now, as I don't see them exposed as mixins in the ds6/mixins. For example, how would the skin page notice integrate the close background icon?

Comment (RajaRamu) [reply dist/less/ds6/mixins.less]: @ianmcburnie To use it in `skin` internally, this is what i am thinking. please correct me if i am wrong,
Existing:
```
.page-notice__close {
    border: 0;
    background: none;
    width: 50px;
    height: 48px;
    cursor: pointer;
    span {
        .icon-close;
    }
}
```
will be chagned to,
```
.page-notice__close {
    border: 0;
    background: none;
    width: 50px;
    height: 48px;
    cursor: pointer;
}
```
Existing Template:
```
<section aria-labelledby="information-status" class="page-notice page-notice--information" role="region">
    <h5 class="page-notice__status">
        <span aria-label="Information notice" id="information-status" role="img"></span>
    </h5>
    <span class="page-notice__cell page-notice__cell--align-middle">
        <p>Try refining your search term for more results.</p>
    </span>
    <button aria-label="Dismiss notice" class="page-notice__close" type="button">
        <span aria-hidden="true" role="img"></span>
    </button>
</section>
```
will be changed to 
```
<section aria-labelledby="information-status" class="page-notice page-notice--information" role="region">
    <h5 class="page-notice__status">
        <span aria-label="Information notice" id="information-status" role="img"></span>
    </h5>
    <span class="page-notice__cell page-notice__cell--align-middle">
        <p>Try refining your search term for more results.</p>
    </span>
    <button aria-label="Dismiss notice" class="page-notice__close" type="button">
        <span class='icon icon-close'></span>
    </button>
</section>
```

Comment (ianmcburnie) [reply dist/less/ds6/mixins.less]: This is incorrect. As I explained in a previous comment, the page notice should not need need to specify `icon icon-close` on the span. Because we know it is always the close icon we can take care of that detail ourselves. Like so:

```less
.page-notice__close span {
  .icon-close-mixin;
}
```

Now no need to specify `icon icon--close`.

```html
<button aria-label="Dismiss notice" class="page-notice__close" type="button">
  <span></span>
</button>
```

Comment (seangates) [reply dist/less/ds6/mixins.less]: @RajaRamu Changing the HTML contract doesn't make any sense. The whole idea is to make it _easier_ for developers. This just makes them have to do more.

Additionally, you've removed the necessary accessibility of the icon.

You need to keep things the way they are, and provide mixins as Ian has suggested.

Comment (RajaRamu) [reply dist/less/ds6/mixins.less]: @ianmcburnie @seangates - makes sense. i will expose it as a mixin for background svg. how about inline svg. how do deal with <symbol> and <svg>?. should it be exposed as mixin too?.

**Comment by ianmcburnie** [comment]
Only if every inline SVG needs some specifically tailored CSS. Like width and height for example.

File: dist/less/ds6/mixins.less (hunk 4 above)
Comment (ianmcburnie) [reply dist/less/ds6/mixins.less]: Only if every inline SVG needs some specifically tailored CSS. Like width and height for example.

**Comment by ianmcburnie** [comment]
screen shot 2018-04-12 at 1 28 15 pm

I still see one of the old icons remaining on the far left. This is the one currently being used in CTA buttons. Please remove from here and use your new icon in Button CTA. This will help verify that your icons are working correctly.

**Comment by ianmcburnie** [comment]
@RajaRamu Can you please double check the dimensions of all the background icons. I was under the impression that the span would exactly match the dimensions of the graphic:

See here that is not the case:

screen shot 2018-04-12 at 1 41 46 pm

**Comment by RajaRamu** [comment]
@ianmcburnie  - sizes match with container size in zeplin - https://app.zeplin.io/project/5a742fd1f51a76f8651f2016/screen/5ac29323882d1f0d438f51ab

zeplin

**Comment by ianmcburnie** [comment]
@RajaRamu That may be true, but there is clearly additional space around the icon in inspector, so something is not right. Did you use the cropped version of the asset from Joe (i.e. the one with no bounding box)? Please investigate and remove this additional space around icon because we need a tight crop in order to exactly control CSS margin and padding.

**Comment by ianmcburnie** [comment]
Would really like us to get some progress on this PR so that we may move forward with 4.0.0 release. 

@RajaRamu This PR also needs to demonstrate how the icons will be used internally by skin.

Example 1: the background SVG for page notice close

Example 2: the inline SVG for CTA button 

If those two work then in another PR we can go and address the other instances of icon usage (dialog, pagination, etc).

**Comment by ianmcburnie** [comment]
After chatting with Raja, I believe the whitespace is because the paths were encoded from a crop of an SVG with 16x16 artboard, but the CSS width and height here has been based on a crop from a 24x24 Zeplin artboard. Going to re-encode the SVG paths based on a new crop from 24x24 artboard. Then things should be in sync.

**Comment by ianmcburnie** [comment]
After looking in more detail, I also notice the `background-size` property is missing. That is the culprit of the white space.

But after re-cropping and re-encoding from a 24x24 source, I notice that some of the current width and height values need slight modification too (a pixel or two out in some cases). I will take care of this.

**Comment by ianmcburnie** [comment]
So, after further investigation, all inline SVGs need their paths and viewboxes updating. This is because the current paths were based on the cropped versions of assets. As previously discussed we want the inline SVGs to be based on the non-cropped version of assets and then use viewbox attribute to do cropping. I have cloned this branch in which I will continue this task. I will close this PR and create a new one from new branch when ready.
